src/AutoMapper/PropertyMap.cs

```diff
@@ -195,117 +199,137 @@ internal void Seal(TypeMapRegistry typeMapRegistry)
         }
 
-        private Func<object, ResolutionContext, object> BuildValueResolverFunc()
+        private Expression<Func<object, ResolutionContext, object>> BuildValueResolverFunc(TypeMapRegistry typeMapRegistry)
         {
-            Func<ResolutionContext, object> valueResolverFunc;
-            
+            Expression valueResolverFunc;
+
+            var srcParam = Expression.Parameter(typeof(object), "src");
+            var ctxtParam = Expression.Parameter(typeof(ResolutionContext), "ctxt");
+
             if (ValueResolverConfig != null)
             {
-
-                Func<ResolutionContext, IValueResolver> ctor;
+                Expression ctor;
                 if (ValueResolverConfig.Instance != null)
                 {
-                    ctor = ctxt => ValueResolverConfig.Instance;
+                    //ctor = ctxt => ValueResolverConfig.Instance;
+                    ctor = Expression.Constant(ValueResolverConfig.Instance);
                 }
                 else if (ValueResolverConfig.Constructor != null)
                 {
-                    ctor = ctxt => ValueResolverConfig.Constructor();
+                    //ctor = ctxt => ValueResolverConfig.Constructor();
+                    ctor = Expression.Invoke(Expression.Constant(ValueResolverConfig.Constructor));
                 }
                 else
                 {
-                    ctor = ctxt => (IValueResolver) ctxt.Options.ServiceCtor(ValueResolverConfig.Type);
+                    //ctor = ctxt => (IValueResolver) ctxt.Options.ServiceCtor(ValueResolverConfig.Type);
+                    ctor = Expression.Convert(
+                        Expression.Invoke(
+                            Expression.MakeMemberAccess(
+                                Expression.MakeMemberAccess(ctxtParam, typeof (ResolutionContext).GetProperty("Options")),
+                                typeof (MappingOperationOptions).GetProperty("ServiceCtor"))
+                            , Expression.Constant(ValueResolverConfig.Type)),
+                        typeof (IValueResolver)
+                        );
                 }
 
-                Func<ResolutionContext, object> sourceFunc;
+                Expression sourceFunc;
                 if (ValueResolverConfig.SourceMember != null)
                 {
-                    var newParam = Expression.Parameter(typeof (object), "m");
-                    var expr = new ConvertingVisitor(ValueResolverConfig.SourceMember.Parameters[0], newParam).Visit(ValueResolverConfig.SourceMember.Body);
-                    var lambda = Expression.Lambda<Func<object, object>>(Expression.Convert(expr, typeof (object)), newParam);
-                    Func<object, object> mapFunc = lambda.Compile();
-
-                    sourceFunc = ctxt => mapFunc(ctxt.SourceValue);
+                    sourceFunc = new ConvertingVisitor(ValueResolverConfig.SourceMember.Parameters[0], srcParam).Visit(ValueResolverConfig.SourceMember.Body);

```

Comment:
I would suggest making an extension method for this.  It's probably going to be used quite a bit with all the expression translations
```
public static Expression ReplaceParameters(this LamdbaExpression lambda, param Expression[] replacements)
{
    Expression convertedExpression = lambda.Body;
    for(var i = 0;i < replacments.Length;i++)
        convertedExpression = new ConvertingVisitor(lambda.Parameters[i], replacements[i]).Visit(convertedExpression);
    return convertedExpression;
}
```